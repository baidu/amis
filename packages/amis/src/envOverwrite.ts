/**
 * @file 用于在移动端或不同语言环境下使用不同配置
 */

import {SchemaNode, Schema} from './types';
import {RendererProps, RendererConfig, addSchemaFilter} from './factory';
import {findObjectsWithKey} from './utils/helper';
const isMobile = (window as any).matchMedia?.('(max-width: 768px)').matches
  ? true
  : false;

// 这里不能用 addSchemaFilter 是因为还需要更深层的替换，比如 select 里的 options
export const envOverwrite = (schema: any, locale?: string) => {
  if (schema.mobile && isMobile) {
    Object.assign(schema, schema.mobile);
    delete schema.mobile;
  }

  if (locale) {
    let schemaNodes = findObjectsWithKey(schema, locale);
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode[locale]);
      delete schemaNode[locale];
    }
  }

  if (isMobile) {
    let schemaNodes = findObjectsWithKey(schema, 'mobile');
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode['mobile']);
      delete schemaNode['mobile'];
    }
  }
};
