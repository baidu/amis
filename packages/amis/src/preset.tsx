import {
  addRootWrapper,
  extendDefaultEnv,
  LazyComponent,
  render,
  themeable,
  ThemeProps
} from 'amis-core';

import {alert, confirm, setRenderSchemaFn, toast, ImageGallery} from 'amis-ui';
import React from 'react';

extendDefaultEnv({
  alert,
  confirm,
  notify: (type: keyof typeof toast, msg: string, conf: any) =>
    toast[type] ? toast[type](msg, conf) : console.warn('[Notify]', type, msg)
});

setRenderSchemaFn((controls, value, callback, scopeRef, theme) => {
  return render(
    {
      name: 'form',
      type: 'form',
      wrapWithPanel: false,
      mode: 'horizontal',
      controls,
      messages: {
        validateFailed: ''
      }
    },
    {
      data: value,
      onFinished: callback,
      scopeRef,
      theme
    },
    {
      session: 'prompt'
    }
  );
});

addRootWrapper((props: any) => {
  const {env, children} = props;
  return (
    <ImageGallery modalContainer={env.getModalContainer}>
      {children}
    </ImageGallery>
  );
});

const SimpleSpinner = themeable(
  (
    props: {
      className?: string;
      spinnerClassName?: string;
    } & ThemeProps
  ) => {
    const cx = props.classnames;

    return (
      <div
        data-testid="spinner"
        className={cx(`Spinner`, 'in', props.className)}
      >
        <div
          className={cx(
            `Spinner-icon`,
            'Spinner-icon--default',
            props.spinnerClassName
          )}
        ></div>
      </div>
    );
  }
);

LazyComponent.defaultProps.placeholder = <SimpleSpinner />;
