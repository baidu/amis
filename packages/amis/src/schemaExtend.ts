/**
 * 扩展 Schema，目前用于实现 input-kv
 */
import isEqual from 'lodash/isEqual';
import {Schema} from 'amis-core';
import {addSchemaFilter} from 'amis-core';
import {isObject} from 'amis-core';

// input-kv 实际上是 combo 的一种扩展
addSchemaFilter(function (schema: Schema, renderer, props?: any) {
  if (schema && schema.type === 'input-kv') {
    const autoParseJSON = schema.autoParseJSON ?? true;
    return {
      draggable: true,
      ...schema,
      multiple: true,
      pipeIn: function (this: any, value: any) {
        if (!isObject(value)) {
          return [];
        }
        if (isEqual(value, this.cachedValue)) {
          return this.cachedValueArray;
        }
        const arr: Array<any> = [];
        Object.keys(value).forEach(key => {
          const valueType = typeof value[key];
          arr.push({
            key: key || '',
            value:
              valueType === 'string' ||
              valueType === 'number' ||
              valueType === 'boolean'
                ? value[key]
                : JSON.stringify(value[key])
          });
        });
        return arr;
      },
      pipeOut: function (this: any, value: any) {
        if (!Array.isArray(value)) {
          return value;
        }
        const obj: any = {};
        value.forEach((item: any) => {
          const key: string = item.key ?? '';
          let value: any = item.value ?? schema.defaultValue ?? '';
          if (
            autoParseJSON &&
            typeof value === 'string' &&
            value.startsWith('{')
          ) {
            try {
              value = JSON.parse(value);
            } catch (e) {}
          }

          obj[key] = value;
        });
        this.cachedValue = obj;
        this.cachedValueArray = value;
        return obj;
      },
      items: [
        schema.keySchema
          ? {
              ...schema.keySchema,
              unique: true,
              name: 'key',
              required: true,
              validateOnChange: true
            }
          : {
              placeholder: schema.keyPlaceholder ?? 'Key',
              type: 'input-text',
              unique: true,
              name: 'key',
              required: true,
              validateOnChange: true
            },
        schema.valueSchema
          ? {...schema.valueSchema, name: 'value'}
          : schema.valueComponent
          ? {
              placeholder: schema.valuePlaceholder ?? 'Value',
              component: schema.valueComponent,
              asFormItem: true,
              name: 'value'
            }
          : {
              placeholder: schema.valuePlaceholder ?? 'Value',
              type: schema.valueType || 'input-text',
              name: 'value'
            }
      ]
    };
  }

  return schema;
});

/**
 * 之前 input-kv 的 value 值不支持对象或数组
 * 很多属性是给单个值设置的，比如 valuePlaceholder 导致
 * 耦合在一起会导致配置项混乱，所以新增了这个组件专门支持 value 是对象或数组的场景
 */
addSchemaFilter(function (schema: Schema, renderer, props?: any) {
  if (schema && schema.type === 'input-kvs') {
    const keyItem = schema.keyItem || {};
    const valueItems = schema.valueItems || [];
    // value 直接放在 key 下的情况
    let flatValue = false;
    if (valueItems.length == 1) {
      if (valueItems[0].name === '_value') {
        flatValue = true;
      }
    }
    const newSchema = {
      draggable: true,
      multiple: true,
      multiLine: true,
      ...schema,
      pipeIn: (data: any) => {
        if (!isObject(data)) {
          return [];
        }
        const arr: Array<any> = [];
        Object.keys(data).forEach(key => {
          let value = data[key];
          if (!value) {
            // value = {};
          }
          if (flatValue) {
            arr.push({
              _key: key || '',
              _value: value
            });
          } else if (typeof value === 'object') {
            arr.push({
              ...value,
              _key: key || ''
            });
          }
        });
        return arr;
      },
      pipeOut: (value: any) => {
        if (!Array.isArray(value)) {
          return value;
        }
        const obj: any = {};
        value.forEach((item: any) => {
          let {_key, ...rest} = item;
          _key = _key ?? '';
          if (flatValue) {
            if (schema.valueIsArray) {
              obj[_key] = item['_value'] || [];
            } else {
              obj[_key] = item['_value'] || {};
            }
            // 数组的时候初始化会生成 [{}]，还不确定是哪生成的，先修正为 []
            if (isEqual(obj[_key], [{}])) {
              obj[_key] = [];
            }
          } else {
            if (schema.valueIsArray) {
              obj[_key] = rest || [];
            } else {
              obj[_key] = rest || {};
            }
          }
        });
        return obj;
      },
      items: [
        {
          type: 'input-text',
          unique: true,
          name: '_key',
          required: true,
          validateOnChange: true,
          ...keyItem
        },
        ...valueItems
      ]
    };
    return newSchema;
  }

  return schema;
});
