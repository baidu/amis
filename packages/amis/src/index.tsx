/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
export * from 'amis-core';
export * from 'amis-ui';

import './preset';

// 注册渲染器
import './renderers/Action';
import './renderers/Alert';
import './renderers/App';
import './renderers/Avatar';
import './renderers/Remark';
import './renderers/ButtonGroup';
import './renderers/Form/ButtonToolbar';
import './renderers/Breadcrumb';
import './renderers/DropDownButton';
import './renderers/Calendar';
import './renderers/Collapse';
import './renderers/CollapseGroup';
import './renderers/Color';
import './renderers/CRUD';
import './renderers/CRUD2';
import './renderers/Pagination';
import './renderers/Cards';
import './renderers/Card';
import './renderers/Card2';
import './renderers/Custom';
import './renderers/Date';
import './renderers/Dialog';
import './renderers/Divider';
import './renderers/Each';
import './renderers/Flex';
import './renderers/Form/Control';
import './renderers/Form/Hidden';
import './renderers/Form/InputText';
import './renderers/Form/InputTag';
import './renderers/Form/InputNumber';
import './renderers/Form/Textarea';
import './renderers/Form/Checkboxes';
import './renderers/Form/Checkbox';
import './renderers/Form/InputCity';
import './renderers/Form/ChartRadios';
import './renderers/Form/InputRating';
import './renderers/Form/Switch';
import './renderers/Form/Radios';
import './renderers/Form/Radio';
import './renderers/Form/JSONSchema';
import './renderers/Form/JSONSchemaEditor';
import './renderers/Form/ListSelect';
import './renderers/Form/LocationPicker';
import './renderers/Form/Select';
import './renderers/Form/Static';
import './renderers/Form/InputDate';
import './renderers/Form/InputDateRange';
import './renderers/Form/InputFormula';
import './renderers/Form/InputRepeat';
import './renderers/Form/InputTree';
import './renderers/Form/TreeSelect';
import './renderers/Form/InputImage';
import './renderers/Form/InputFile';
import './renderers/Form/UUID';
import './renderers/Form/MatrixCheckboxes';
import './renderers/Form/InputMonthRange';
import './renderers/Form/InputQuarterRange';
import './renderers/Form/InputYearRange';
import './renderers/Form/InputRange';
import './renderers/Form/InputArray';
import './renderers/Form/Combo';
import './renderers/Form/ConditionBuilder';
import './renderers/Form/InputSubForm';
import './renderers/Form/InputExcel';
import './renderers/Form/InputRichText';
import './renderers/Form/Editor';
import './renderers/Form/DiffEditor';
import './renderers/Form/InputColor';
import './renderers/Form/ChainedSelect';
import './renderers/Form/NestedSelect';
import './renderers/Form/Transfer';
import './renderers/Form/TransferPicker';
import './renderers/Form/InputTable';
import './renderers/Form/Picker';
import './renderers/Form/IconPicker';
import './renderers/Form/IconSelect';
import './renderers/Form/Formula';
import './renderers/Form/FieldSet';
import './renderers/Form/TabsTransfer';
import './renderers/Form/TabsTransferPicker';
import './renderers/Form/Group';
import './renderers/Form/InputGroup';
import './renderers/Form/UserSelect';
import './renderers/Grid';
import './renderers/Grid2D';
import './renderers/HBox';
import './renderers/VBox';
import './renderers/Image';
import './renderers/Images';
import './renderers/List';
import './renderers/Log';
import './renderers/Operation';
import './renderers/Page';
import './renderers/PaginationWrapper';
import './renderers/Panel';
import './renderers/Plain';
import './renderers/Property';
import './renderers/Portlet';
import './renderers/Spinner';
import './renderers/Table/index';
import './renderers/Tabs';
import './renderers/Tpl';
import './renderers/Mapping';
import './renderers/Progress';
import './renderers/Status';
import './renderers/Json';
import './renderers/Link';
import './renderers/Wizard';
import './renderers/Chart';
import './renderers/Container';
import './renderers/SwitchContainer';
import './renderers/SearchBox';
import './renderers/Service';
import './renderers/SparkLine';
import './renderers/Video';
import './renderers/Audio';
import './renderers/Nav';
import './renderers/Number';
import './renderers/Tasks';
import './renderers/Drawer';
import './renderers/Wrapper';
import './renderers/IFrame';
import './renderers/BarCode';
import './renderers/QRCode';
import './renderers/Icon';
import './renderers/Carousel';
import './renderers/AnchorNav';
import './renderers/Steps';
import './renderers/Timeline';
import './renderers/Markdown';
import './renderers/TableView';
import './renderers/Code';
import './renderers/WebComponent';
import './renderers/GridNav';
import './renderers/TooltipWrapper';
import './renderers/Tag';
import './renderers/Table2/index';
import './renderers/Words';
import './renderers/Password';
import './renderers/DateRange';
import './renderers/MultilineText';
import './renderers/OfficeViewer';
import './renderers/AMIS';

import './compat';
import './schemaExtend';
import type {
  BaseSchema,
  FormSchema,
  SchemaApi,
  SchemaCollection,
  SchemaExpression,
  SchemaObject
} from './Schema';
import type {TableViewSchema, TrObject} from './renderers/TableView';
import type {ActionSchema, ButtonSchema} from './renderers/Action';
import type {CRUDCommonSchema} from './renderers/CRUD';
import type {CRUD2Schema} from './renderers/CRUD2';
import type {TabsSchema} from './renderers/Tabs';
import {availableLanguages as EditorAvailableLanguages} from './renderers/Form/Editor';
import type {Action} from './types';
export * from './renderers/Form/IconPickerIcons';
export * from './renderers/Form/IconSelectStore';

export {
  BaseSchema,
  SchemaCollection,
  FormSchema,
  SchemaApi,
  SchemaObject,
  TableViewSchema,
  TrObject,
  ActionSchema,
  CRUDCommonSchema,
  ButtonSchema,
  CRUD2Schema,
  TabsSchema,
  SchemaExpression,
  Action,
  EditorAvailableLanguages
};
