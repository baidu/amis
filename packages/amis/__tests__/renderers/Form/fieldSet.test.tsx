/**
 * 组件名称：FieldSet 表单项集合
 *
 * 单测内容：
 * 1. 基本使用
 * 2. mode
 * 3. 可折叠 collapsable 和 默认是否折叠 collapsed
 * 4. 标题位置 titlePosition、 收起标题 collapseTitle、展开标题 title
 * 5. size
 */

import {render, fireEvent} from '@testing-library/react';
import '../../../src';
import {render as amisRender} from '../../../src';
import {makeEnv, wait} from '../../helper';

// 1. 基本使用
test('Renderer:fieldSet', async () => {
  const {container} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        controls: [
          {
            type: 'fieldSet',
            name: 'a',
            label: 'fieldSet',
            mode: 'horizontal',
            collapsable: true,
            collapsed: false,
            className: 'no-border',
            headingClassName: 'bg-dark',
            bodyClassName: 'bg-white',
            horizontal: {
              leftFixed: 1,
              left: 4,
              right: 7
            },
            controls: [
              {
                name: 'text',
                type: 'text'
              }
            ]
          }
        ],
        title: 'The form',
        actions: []
      },
      {},
      makeEnv({})
    )
  );

  expect(container).toMatchSnapshot();
});

// 2. mode
test('Renderer:fieldSet with mode', async () => {
  const schema = (mode: any) => ({
    type: 'form',
    api: '/api/xxx',
    body: [
      {
        type: 'fieldSet',
        name: 'a',
        label: 'fieldSet',
        mode,
        body: [
          {
            name: 'text1',
            type: 'input-text',
            label: '文本1'
          },
          {
            name: 'text2',
            type: 'input-text',
            label: '文本2'
          }
        ]
      }
    ],
    title: 'The form',
    actions: []
  });
  const {container, rerender} = render(
    amisRender(schema('horizontal'), {}, makeEnv({}))
  );
  expect(container.querySelector('.cxd-Form--horizontal')!).toBeInTheDocument();
  expect(container).toMatchSnapshot();

  rerender(amisRender(schema('inline'), {}, makeEnv({})));
  expect(container.querySelector('.cxd-Form--inline')!).toBeInTheDocument();
  expect(container).toMatchSnapshot();

  rerender(amisRender(schema('justify'), {}, makeEnv({})));
  expect(container.querySelector('.cxd-Form--justify')!).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

// 3. 可折叠 collapsable 和 默认是否折叠 collapsed
test('Renderer:fieldSet with collapsable & collapsed', async () => {
  const {container, rerender} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        body: [
          {
            type: 'fieldSet',
            name: 'a',
            label: 'fieldSet',
            collapsable: true,
            title: 'header',
            body: [
              {
                name: 'text1',
                type: 'input-text',
                label: '文本1'
              },
              {
                name: 'text2',
                type: 'input-text',
                label: '文本2'
              }
            ]
          },
          {
            type: 'fieldSet',
            name: 'b',
            label: 'fieldSet2',
            collapsable: true,
            collapsed: true,
            title: 'header',
            body: [
              {
                name: 'text1',
                type: 'input-text',
                label: '文本1'
              },
              {
                name: 'text2',
                type: 'input-text',
                label: '文本2'
              }
            ]
          }
        ],
        title: 'The form',
        actions: []
      },
      {},
      makeEnv({})
    )
  );

  const header = container.querySelector('.cxd-Collapse-header')!;
  expect(container.querySelector('.cxd-Collapse')!).toHaveClass('is-active');
  expect(
    container.querySelector('.cxd-Collapse:last-of-type')!
  ).not.toHaveClass('is-active');

  fireEvent.click(header);
  await wait(300);
  expect(container.querySelector('.cxd-Collapse')!).not.toHaveClass(
    'is-active'
  );
});

// 4. 标题位置 titlePosition、 收起标题 collapseTitle、展开标题 title
test('Renderer:fieldSet with titlePosition & collapseTitle & title', async () => {
  const {container, rerender} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        body: [
          {
            type: 'fieldSet',
            name: 'a',
            label: 'fieldSet',
            collapsable: true,
            title: '展开更多设置',
            collapseTitle: '收起设置',
            titlePosition: 'bottom',
            body: [
              {
                name: 'text1',
                type: 'input-text',
                label: '文本1'
              },
              {
                name: 'text2',
                type: 'input-text',
                label: '文本2'
              }
            ]
          }
        ],
        title: 'The form',
        actions: []
      },
      {},
      makeEnv({})
    )
  );

  const item = container.querySelector('.cxd-Collapse.is-active')!;

  expect(item).toHaveClass('cxd-Collapse--title-bottom');
  expect(item.firstElementChild).toHaveClass('cxd-Collapse-contentWrapper');
  expect(item.lastElementChild).toHaveTextContent('收起设置');
  expect(container).toMatchSnapshot();

  fireEvent.click(item.lastElementChild!);
  await wait(400);

  expect(item.lastElementChild).toHaveTextContent('展开更多设置');
});

// 5. size
test('Renderer:fieldSet with titlePosition & collapseTitle & title', async () => {
  const schema = (size: any) => ({
    type: 'form',
    api: '/api/xxx',
    body: [
      {
        type: 'fieldSet',
        name: 'a',
        label: 'fieldSet',
        collapsable: true,
        title: '展开更多设置',
        size,
        body: [
          {
            name: 'text1',
            type: 'input-text',
            label: '文本1'
          },
          {
            name: 'text2',
            type: 'input-text',
            label: '文本2'
          }
        ]
      }
    ],
    title: 'The form',
    actions: []
  });
  const {container, rerender} = render(
    amisRender(schema('xs'), {}, makeEnv({}))
  );
  expect(
    container.querySelector('.cxd-Collapse.cxd-Collapse--xs')!
  ).toBeInTheDocument();

  rerender(amisRender(schema('sm'), {}, makeEnv({})));
  expect(
    container.querySelector('.cxd-Collapse.cxd-Collapse--sm')!
  ).toBeInTheDocument();

  rerender(amisRender(schema('base'), {}, makeEnv({})));
  expect(
    container.querySelector('.cxd-Collapse.cxd-Collapse--base')!
  ).toBeInTheDocument();

  rerender(amisRender(schema('lg'), {}, makeEnv({})));
  expect(
    container.querySelector('.cxd-Collapse.cxd-Collapse--lg')!
  ).toBeInTheDocument();

  rerender(amisRender(schema('xl'), {}, makeEnv({})));
  expect(
    container.querySelector('.cxd-Collapse.cxd-Collapse--xl')!
  ).toBeInTheDocument();
});
