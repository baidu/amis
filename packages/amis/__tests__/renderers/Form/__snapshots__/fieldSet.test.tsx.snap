// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:fieldSet 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <fieldset
          class="cxd-Collapse is-active no-border"
        >
          <div
            class="cxd-Collapse-contentWrapper"
          >
            <div
              class="cxd-Collapse-body bg-white"
            >
              <div
                class="cxd-Collapse-content"
              >
                <div
                  class="cxd-Form--horizontal"
                >
                  <div
                    class="cxd-Form-item cxd-Form-item--horizontal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label cxd-Form-itemColumn--normal"
                    >
                      <span />
                    </label>
                    <div
                      class="cxd-Form-value"
                    >
                      <div
                        class="cxd-Form-control cxd-TextControl"
                      >
                        <div
                          class="cxd-TextControl-input"
                        >
                          <input
                            autocomplete="off"
                            class=""
                            name="text"
                            placeholder=""
                            size="10"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</div>
`;

exports[`Renderer:fieldSet with mode 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <fieldset
          class="cxd-Collapse is-active"
        >
          <div
            class="cxd-Collapse-contentWrapper"
          >
            <div
              class="cxd-Collapse-body"
            >
              <div
                class="cxd-Collapse-content"
              >
                <div
                  class="cxd-Form--horizontal"
                >
                  <div
                    class="cxd-Form-item cxd-Form-item--horizontal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label cxd-Form-itemColumn--normal"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本1
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-value"
                    >
                      <div
                        class="cxd-Form-control cxd-TextControl"
                      >
                        <div
                          class="cxd-TextControl-input"
                        >
                          <input
                            autocomplete="off"
                            class=""
                            name="text1"
                            placeholder=""
                            size="10"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="cxd-Form-item cxd-Form-item--horizontal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label cxd-Form-itemColumn--normal"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本2
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-value"
                    >
                      <div
                        class="cxd-Form-control cxd-TextControl"
                      >
                        <div
                          class="cxd-TextControl-input"
                        >
                          <input
                            autocomplete="off"
                            class=""
                            name="text2"
                            placeholder=""
                            size="10"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</div>
`;

exports[`Renderer:fieldSet with mode 2`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <fieldset
          class="cxd-Collapse is-active"
        >
          <div
            class="cxd-Collapse-contentWrapper"
          >
            <div
              class="cxd-Collapse-body"
            >
              <div
                class="cxd-Collapse-content"
              >
                <div
                  class="cxd-Form--inline"
                >
                  <div
                    class="cxd-Form-item cxd-Form-item--inline"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本1
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-value"
                    >
                      <div
                        class="cxd-Form-control cxd-TextControl"
                      >
                        <div
                          class="cxd-TextControl-input"
                        >
                          <input
                            autocomplete="off"
                            class=""
                            name="text1"
                            placeholder=""
                            size="10"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="cxd-Form-item cxd-Form-item--inline"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本2
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-value"
                    >
                      <div
                        class="cxd-Form-control cxd-TextControl"
                      >
                        <div
                          class="cxd-TextControl-input"
                        >
                          <input
                            autocomplete="off"
                            class=""
                            name="text2"
                            placeholder=""
                            size="10"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</div>
`;

exports[`Renderer:fieldSet with mode 3`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <fieldset
          class="cxd-Collapse is-active"
        >
          <div
            class="cxd-Collapse-contentWrapper"
          >
            <div
              class="cxd-Collapse-body"
            >
              <div
                class="cxd-Collapse-content"
              >
                <div
                  class="cxd-Form--justify"
                >
                  <div
                    class="cxd-Form-item cxd-Form-item--normal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本1
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-control cxd-TextControl"
                    >
                      <div
                        class="cxd-TextControl-input"
                      >
                        <input
                          autocomplete="off"
                          class=""
                          name="text1"
                          placeholder=""
                          size="10"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="cxd-Form-item cxd-Form-item--normal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本2
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-control cxd-TextControl"
                    >
                      <div
                        class="cxd-TextControl-input"
                      >
                        <input
                          autocomplete="off"
                          class=""
                          name="text2"
                          placeholder=""
                          size="10"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</div>
`;

exports[`Renderer:fieldSet with titlePosition & collapseTitle & title 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <fieldset
          class="cxd-Collapse is-active cxd-Collapse--title-bottom"
        >
          <div
            class="cxd-Collapse-contentWrapper"
          >
            <div
              class="cxd-Collapse-body"
            >
              <div
                class="cxd-Collapse-content"
              >
                <div
                  class="cxd-Form--normal"
                >
                  <div
                    class="cxd-Form-item cxd-Form-item--normal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本1
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-control cxd-TextControl"
                    >
                      <div
                        class="cxd-TextControl-input"
                      >
                        <input
                          autocomplete="off"
                          class=""
                          name="text1"
                          placeholder=""
                          size="10"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="cxd-Form-item cxd-Form-item--normal"
                    data-role="form-item"
                  >
                    <label
                      class="cxd-Form-label"
                    >
                      <span>
                        <span
                          class="cxd-TplField"
                        >
                          <span>
                            文本2
                          </span>
                        </span>
                      </span>
                    </label>
                    <div
                      class="cxd-Form-control cxd-TextControl"
                    >
                      <div
                        class="cxd-TextControl-input"
                      >
                        <input
                          autocomplete="off"
                          class=""
                          name="text2"
                          placeholder=""
                          size="10"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="cxd-Collapse-header"
          >
            <span
              class="cxd-Collapse-arrow-wrap"
            >
              <icon-mock
                classname="cxd-Collapse-arrow icon icon-right-arrow-bold cxd-Collapse-arrow"
                icon="right-arrow-bold"
              />
            </span>
            <span
              class="cxd-TplField"
            >
              <span>
                收起设置
              </span>
            </span>
          </div>
        </fieldset>
      </form>
    </div>
  </div>
</div>
`;
